eigh() tests fail to pass, crash Python with seemingly ramdom pattern

This problem is related to #11601, which has been closed by #11702 ( @ilayn ). However, the crash has not been fixed by the latter PR.

The symptoms remained almost identical to the one described in my comment in https://github.com/scipy/scipy/issues/11601#issuecomment-600153321

In summary, when running the test for `eigh()`, Python tends to crash with SIGSEGV or SIGABRT. Sometimes this happens during the `test_eigh()` function, sometimes after it passed with "100%" but before `pytest` returns.

The test that triggers the crash is the following test function:

https://github.com/scipy/scipy/blob/ae34ce4835949a8310d7c3d7bcb4a55aafd11f4f/scipy/linalg/tests/test_decomp.py#L863-L888

#### Some patterns from the histories of crashes

I run the test script with `runtests.py` 100 times and saved the output as text files.

By grepping the output files `./runtests.py`, I notice that the last-known position in Python before it crashes could be three lines, namely 873, 876, and 877. L 873 is the actual call to `eigh()`, **while the crash can happen as late as 876 or 877**, where the arrays returned from `eigh()` are accessed.

Only 6 out of 100 runs passed without any problems.

In some cases (35 out of the 100), Python segfaults after nominally completing all the tests in `TestEigh::test_eigh`.

In the cases where Python was killed with SIGABRT, 36 were at L 873 (call to `eigh()`), while 9 were at L 876 where output `z` was used. In many other runs, the test script was not featured in the Python backtrace if any.

The parametrized inputs that triggered the crash were of the form `test_eigh[6-D-XXX-YYY-ZZZ-eigvals1]`. That is, the crashes happened for dimension 6, dtype double complex, with `eigvals=` keyword parameter set to the tuple `(2, 4)`. The `XXX`--`ZZZ` parameters are boolean flags for keywords `turbo`, `lower`, and `overwrite` respectively.

An incomplete tally of the parameters (`turbo`, `lower`, and `overwrite`), where Python crashed before finishing all the tests, is as follows:

```
   5 False-False-False
  11 False-False-True
  13 False-True-False
   6 False-True-True
   7 True-False-False
   4 True-False-True
  15 True-True-True
```

The combination `(turbo=True, lower=True, overwrite=False)` is the one missing from the 2^3 = 8 cases yet.

#### Reproducing code example:

```shell
./runtests.py -vt scipy/linalg/tests/test_decomp.py::TestEigh::test_eigh
```

#### Scipy/Numpy/Python version information:

Scipy master branch as of ae34ce48, Numpy 1.18.1, Python 3.7.6, conda macos with MKL 2019.4.

Note sure if this is helpful but posting: 

```
(scipy-dev) Lucass-MacBook:scipy rlucas$ ./runtests.py -vt scipy/linalg/tests/test_decomp.py::TestEigh::test_eigh

... elided bulk of passing tests output ...
scipy/linalg/tests/test_decomp.py::TestEigh::test_eigh[6-D-False-True-True-None] PASSED                                                                                       [ 89%]
scipy/linalg/tests/test_decomp.py::TestEigh::test_eigh[6-D-False-True-True-eigvals1] FAILED                                                                                   [ 90%]
scipy/linalg/tests/test_decomp.py::TestEigh::test_eigh[6-D-False-True-False-None] PASSED                                                                                      [ 92%]
scipy/linalg/tests/test_decomp.py::TestEigh::test_eigh[6-D-False-True-False-eigvals1] PASSED                                                                                  [ 93%]
scipy/linalg/tests/test_decomp.py::TestEigh::test_eigh[6-D-False-False-True-None] PASSED                                                                                      [ 95%]
scipy/linalg/tests/test_decomp.py::TestEigh::test_eigh[6-D-False-False-True-eigvals1] PASSED                                                                                  [ 96%]
scipy/linalg/tests/test_decomp.py::TestEigh::test_eigh[6-D-False-False-False-None] PASSED                                                                                     [ 98%]
scipy/linalg/tests/test_decomp.py::TestEigh::test_eigh[6-D-False-False-False-eigvals1] PASSED                                                                                 [100%]

===================================================================================== FAILURES ======================================================================================
_________________________________________________________________ TestEigh.test_eigh[6-D-False-True-True-eigvals1] __________________________________________________________________
scipy/linalg/tests/test_decomp.py:876: in test_eigh
    diag_ = diag(z.T.conj() @ a @ z).real
E   ValueError: matmul: Input operand 1 has a mismatch in its core dimension 0, with gufunc signature (n?,k),(k,m?)->(n?,m?) (size -2308328180370374638 is different from 6)
======================================================================== 1 failed, 63 passed in 0.99 seconds ========================================================================
(scipy-dev) Lucass-MacBook:scipy rlucas$ 
```

--- a/scipy/linalg/flapack_sym_herm.pyf.src
+++ b/scipy/linalg/flapack_sym_herm.pyf.src
@@ -757,22 +757,22 @@ subroutine <prefix2>syevr(compute_v,range,lower,n,a,lda,vl,vu,il,iu,abstol,w,z,m
     integer optional,intent(in),check(compute_v==1||compute_v==0):: compute_v = 1
     character optional,intent(in),check(*range=='A'||*range=='V' ||*range=='I') :: range='A'
     integer optional,intent(in),check(lower==0||lower==1) :: lower = 0
-    integer optional,intent(in),check(il>=1) :: il=1
-    integer optional,intent(in),check(iu>=il||iu<=n),depend(il,n) :: iu=n
+    integer optional,intent(in),check(il>=1&&il<=n),depend(n) :: il=1
+    integer optional,intent(in),check(iu>=il&&iu<=n),depend(il,n) :: iu=n
     <ftype2> optional,intent(in) :: vl=0.0
-    <ftype2> optional,intent(in),check(vu>=vl),depend(vl) :: vu=1.0
+    <ftype2> optional,intent(in),check(vu>vl),depend(vl) :: vu=1.0
     <ftype2> intent(in) :: abstol=0.0
-    integer optional,intent(in),depend(n),check(lwork>=1||lwork==-1) :: lwork=max(26*n,1)
-    integer optional,intent(in),depend(n),check(liwork>=1||liwork==-1):: liwork= max(1,10*n)
+    integer optional,intent(in),depend(n),check(lwork>=max(1,26*n)||lwork==-1) :: lwork=max(26*n,1)
+    integer optional,intent(in),depend(n),check(liwork>=max(1,10*n)||liwork==-1):: liwork= max(1,10*n)
 
     integer intent(hide),depend(a) :: n=shape(a,0)
     integer intent(hide),depend(n) :: lda=max(1,n)
-    integer intent(hide),depend(z) :: ldz=max(1,shape(z,0))
+    integer intent(hide),depend(compute_v,n) :: ldz=(compute_v?max(1,n):1)
     <ftype2>  intent(hide),dimension(lwork),depend(lwork) :: work
     integer intent(hide),dimension(liwork),depend(liwork) :: iwork
 
     <ftype2> intent(out),dimension(n),depend(n) :: w
-    <ftype2> intent(out),dimension((compute_v?MAX(0,n):0),(compute_v?(*range=='I'?iu-il+1:MAX(1,n)):0)),depend(n,compute_v,range,iu,il) :: z
+    <ftype2> intent(out),dimension((compute_v?ldz:0),(compute_v?(*range=='I'?iu-il+1:MAX(1,n)):0)),depend(n,compute_v,ldz,range,iu,il) :: z
     integer intent(out) :: m
     ! Only returned if range=='A' or range=='I' and il, iu = 1, n
     integer intent(out),dimension((compute_v?(2*(*range=='A'||(*range=='I' && iu-il+1==n)?n:0)):0)),depend(n,iu,il,compute_v,range) :: isuppz
@@ -827,27 +827,32 @@ subroutine <prefix2c>heevr(compute_v,range,lower,n,a,lda,vl,vu,il,iu,abstol,w,z,
     integer optional,intent(in),check(compute_v==1||compute_v==0):: compute_v = 1
     character optional,intent(in),check(*range=='A'||*range=='V' ||*range=='I') :: range='A'
     integer optional,intent(in),check(lower==0||lower==1) :: lower = 0
-    integer optional,intent(in) :: il=1
-    integer optional,intent(in),depend(n) :: iu=n
+    integer optional,intent(in),check(il>=1&&il<=n),depend(n) :: il=1
+    integer optional,intent(in),check(iu>=il&&iu<=n),depend(il,n) :: iu=n
     <ftype2> optional,intent(in) :: vl=0.0
     <ftype2> optional,intent(in),check(vu>vl),depend(vl) :: vu=1.0
     <ftype2> intent(in) :: abstol=0.0
-    integer optional,intent(in),depend(n),check(lwork>=1||lwork==-1) :: lwork=max(2*n,1)
-    integer optional,intent(in),depend(n),check(lwork>=1||lwork==-1) :: lrwork=max(24*n,1)
-    integer optional,intent(in),depend(n),check(liwork>=1||liwork==-1):: liwork= max(1,10*n)
+    integer optional,intent(in),depend(n),check(lwork>=max(2*n,1)||lwork==-1) :: lwork=max(2*n,1)
+    integer optional,intent(in),depend(n),check(lrwork>=max(24*n,1)||lrwork==-1) :: lrwork=max(24*n,1)
+    integer optional,intent(in),depend(n),check(liwork>=max(1,10*n)||liwork==-1):: liwork= max(1,10*n)
 
     integer intent(hide),depend(a) :: n=shape(a,0)
     integer intent(hide),depend(n) :: lda=max(1,n)
-    integer intent(hide),depend(z) :: ldz=max(1,shape(z,0))
+    integer intent(hide),depend(compute_v,n) :: ldz=(compute_v?max(1,n):1)
     <ftype2c>  intent(hide),dimension(lwork),depend(lwork) :: work
-    <ftype2> intent(hide),dimension(lrwork) :: rwork
+    <ftype2> intent(hide),dimension(lrwork),depend(lrwork) :: rwork
     integer intent(hide),dimension(liwork),depend(liwork) :: iwork
 
     <ftype2> intent(out),dimension(n),depend(n) :: w
-    <ftype2c> intent(out),dimension((compute_v?MAX(0,n):0),(compute_v?(*range=='I'?iu-il+1:MAX(1,n)):0)),depend(n,compute_v,range,iu,il) :: z
+    <ftype2c> intent(out),dimension((compute_v?ldz:0),(compute_v?(*range=='I'?iu-il+1:MAX(1,n)):0)),depend(n,ldz,compute_v,range,iu,il) :: z
     integer intent(out) :: m
-    ! Only returned if range=='A' or range=='I' and il, iu = 1, n
-    integer intent(out),dimension((compute_v?(2*(*range=='A'||(*range=='I' && iu-il+1==n)?n:0)):0)),depend(n,iu,il,range,compute_v) :: isuppz
+    ! According to the NetLib LAPACK reference-implementation the arrray isuppz
+    ! is only "implemented" if range=='A' or range=='I' and il, iu = 1, n.
+    ! The array is nevertheless allocated with a size that is large enough
+    ! as precaution against out-of-bound access. The underlying LAPACK
+    ! implementation may as well use the buffer in any case, even if the
+    ! eigenvectors are not requested.
+    integer intent(out),dimension(2*max(1,n)),depend(n) :: isuppz
     integer intent(out) :: info
 
 end subroutine <prefix2c>heevr

--- a/scipy/linalg/flapack_sym_herm.pyf.src
+++ b/scipy/linalg/flapack_sym_herm.pyf.src
@@ -757,22 +757,22 @@ subroutine <prefix2>syevr(compute_v,range,lower,n,a,lda,vl,vu,il,iu,abstol,w,z,m
     integer optional,intent(in),check(compute_v==1||compute_v==0):: compute_v = 1
     character optional,intent(in),check(*range=='A'||*range=='V' ||*range=='I') :: range='A'
     integer optional,intent(in),check(lower==0||lower==1) :: lower = 0
-    integer optional,intent(in),check(il>=1&&il<=n),depend(n) :: il=1
-    integer optional,intent(in),check(iu>=il&&iu<=n),depend(il,n) :: iu=n
+    integer optional,intent(in),check(il>=1) :: il=1
+    integer optional,intent(in),check(iu>=il||iu<=n),depend(il,n) :: iu=n
     <ftype2> optional,intent(in) :: vl=0.0
-    <ftype2> optional,intent(in),check(vu>vl),depend(vl) :: vu=1.0
+    <ftype2> optional,intent(in),check(vu>=vl),depend(vl) :: vu=1.0
     <ftype2> intent(in) :: abstol=0.0
     integer optional,intent(in),depend(n),check(lwork>=max(1,26*n)||lwork==-1) :: lwork=max(26*n,1)
     integer optional,intent(in),depend(n),check(liwork>=max(1,10*n)||liwork==-1):: liwork= max(1,10*n)
 
     integer intent(hide),depend(a) :: n=shape(a,0)
     integer intent(hide),depend(n) :: lda=max(1,n)
-    integer intent(hide),depend(compute_v,n) :: ldz=(compute_v?max(1,n):1)
+    integer intent(hide),depend(z) :: ldz=max(1,shape(z,0))
     <ftype2>  intent(hide),dimension(lwork),depend(lwork) :: work
     integer intent(hide),dimension(liwork),depend(liwork) :: iwork
 
     <ftype2> intent(out),dimension(n),depend(n) :: w
-    <ftype2> intent(out),dimension((compute_v?ldz:0),(compute_v?(*range=='I'?iu-il+1:MAX(1,n)):0)),depend(n,compute_v,ldz,range,iu,il) :: z
+    <ftype2> intent(out),dimension((compute_v?MAX(0,n):0),(compute_v?(*range=='I'?iu-il+1:MAX(1,n)):0)),depend(n,compute_v,range,iu,il) :: z
     integer intent(out) :: m
     ! Only returned if range=='A' or range=='I' and il, iu = 1, n
     integer intent(out),dimension((compute_v?(2*(*range=='A'||(*range=='I' && iu-il+1==n)?n:0)):0)),depend(n,iu,il,compute_v,range) :: isuppz
@@ -827,8 +827,8 @@ subroutine <prefix2c>heevr(compute_v,range,lower,n,a,lda,vl,vu,il,iu,abstol,w,z,
     integer optional,intent(in),check(compute_v==1||compute_v==0):: compute_v = 1
     character optional,intent(in),check(*range=='A'||*range=='V' ||*range=='I') :: range='A'
     integer optional,intent(in),check(lower==0||lower==1) :: lower = 0
-    integer optional,intent(in),check(il>=1&&il<=n),depend(n) :: il=1
-    integer optional,intent(in),check(iu>=il&&iu<=n),depend(il,n) :: iu=n
+    integer optional,intent(in),check(il>=1) :: il=1
+    integer optional,intent(in),check(iu>=il||iu<=n),depend(il,n) :: iu=n
     <ftype2> optional,intent(in) :: vl=0.0
     <ftype2> optional,intent(in),check(vu>vl),depend(vl) :: vu=1.0
     <ftype2> intent(in) :: abstol=0.0
@@ -838,20 +838,22 @@ subroutine <prefix2c>heevr(compute_v,range,lower,n,a,lda,vl,vu,il,iu,abstol,w,z,
 
     integer intent(hide),depend(a) :: n=shape(a,0)
     integer intent(hide),depend(n) :: lda=max(1,n)
-    integer intent(hide),depend(compute_v,n) :: ldz=(compute_v?max(1,n):1)
+    integer intent(hide),depend(z) :: ldz=max(1,shape(z,0))
     <ftype2c>  intent(hide),dimension(lwork),depend(lwork) :: work
     <ftype2> intent(hide),dimension(lrwork),depend(lrwork) :: rwork
     integer intent(hide),dimension(liwork),depend(liwork) :: iwork
 
     <ftype2> intent(out),dimension(n),depend(n) :: w
-    <ftype2c> intent(out),dimension((compute_v?ldz:0),(compute_v?(*range=='I'?iu-il+1:MAX(1,n)):0)),depend(n,ldz,compute_v,range,iu,il) :: z
+    <ftype2c> intent(out),dimension((compute_v?MAX(0,n):0),(compute_v?(*range=='I'?iu-il+1:MAX(1,n)):0)),depend(n,compute_v,range,iu,il) :: z
     integer intent(out) :: m
-    ! According to the NetLib LAPACK reference-implementation the arrray isuppz
-    ! is only "implemented" if range=='A' or range=='I' and il, iu = 1, n.
-    ! The array is nevertheless allocated with a size that is large enough
-    ! as precaution against out-of-bound access. The underlying LAPACK
-    ! implementation may as well use the buffer in any case, even if the
-    ! eigenvectors are not requested.
+    ! MKL implementation has a bug that still accesses isuppz array even if
+    ! range=='A' or range=='I' and il, iu = 1, n which is not the case for
+    ! the reference implementation. Hence here isuppz is allocated regardless
+    ! of the settings. It is wasteful but necessary. The bug is fixed in
+    ! mkl 2020 update 2 and when time comes change this line with
+    !
+    ! integer intent(out),dimension((compute_v?(2*(*range=='A'||(*range=='I' && iu-il+1==n)?n:0)):0)),depend(n,iu,il,range,compute_v) :: isuppz
+    !
     integer intent(out),dimension(2*max(1,n)),depend(n) :: isuppz
     integer intent(out) :: info
 

